torch_np/_wrapper.py

```diff
@@ -258,13 +268,60 @@ def corrcoef(x, y=None, rowvar=True, bias=NoValue, ddof=NoValue, *, dtype=None):
 
 
 def concatenate(ar_tuple, axis=0, out=None, dtype=None, casting="same_kind"):
-    if casting != "same_kind":
-        raise NotImplementedError
-    tensors = tuple(asarray(ar).get() for ar in ar_tuple)
-    if dtype is not None:
-        # XXX: map numpy dtypes
-        tensors = tuple(ar.type(dtype) for ar in ar_typle)
-    return asarray(torch.cat(tensors, axis, out=out))
+    if out is not None:
+        if dtype is not None:
+            # mimic numpy
+            raise TypeError("concatenate() only takes `out` or `dtype` as an "
+                            "argument, but both were provided.")
+        if not isinstance(out, ndarray):
+            raise ValueError("'out' must be an array")
+        out_tensor = out.get()
+    else:
+        out_tensor = out
+
+    # make sure inputs are arrays
+    arrays = tuple(asarray(ar) for ar in ar_tuple)
+    tensors = tuple(ar.get() for ar in arrays)
+
+    if arrays == ():
+        raise ValueError("need at least one array to concatenate")
+
+    # XXX: factor out this logic
+    # figure out the type of the inputs and outputs
+    if out is None and dtype is None:
+        out_dtype = None
+    else:
+        out_dtype = _dtypes.dtype(dtype) if dtype is not None else out.dtype
+
+        # check if we can cast all arguments
+        for ar in arrays:
+            if not can_cast(ar.dtype, out_dtype, casting=casting):
+                raise TypeError(f"Cannot cast array data from {ar.dtype} to"
+                                 " {out_dtype} according to the rule '{casting}'")
+
+        # cast
+        torch_dtype = _dtypes.torch_dtype_from(out_dtype)
+        tensors = tuple(tensor.to(torch_dtype) for tensor in tensors)
+    
+    if axis is None:
+        tensors = tuple(tensor.ravel() for tensor in tensors)
+        axis = 0
+
+    try:
+        try:
+            with warnings.catch_warnings():
+                warnings.simplefilter('error', UserWarning)
+                # torch emits a UserWarning if the out tensor has wrong size,
+                # while numpy errors out
+                # TODO: centralize validation of out (incl this numpy/torch discrepancy)
+                result = torch.cat(tensors, axis, out=out_tensor)
+        except UserWarning:
+            raise ValueError("output array has wrong shape or dimensionality")
+    except (IndexError, RuntimeError):
+        raise AxisError
+    if out is not None:
+        return out
+    return asarray(result)
```

Comment:
This `concatenate` functions is a prototype for a top-level namespace function without a corresponding ndarray method.